include/boost/hana/config.hpp

```diff
@@ -70,7 +70,7 @@ Distributed under the Boost Software License, Version 1.0.
 //////////////////////////////////////////////////////////////////////////////
 // Check the compiler for general C++14 capabilities
 //////////////////////////////////////////////////////////////////////////////
-#if (__cplusplus < 201400)
+#if (__cplusplus < 201400) && (defined(__clang__) || !defined(_MSC_VER))
```

Comment:
Why not simply `#if (__cplusplus < 201400) && !defined(_MSC_VER)`? And when `#warning` is not supported, we should fallback to `#pragma message` instead. Giving this message is important to avoid users being totally confused by meaningless error messages when trying to use Hana with a non-compliant compiler.